File: easy-random-bean-validation/src/test/java/org/jeasy/random/validation/FutureOrPresentAnnotationHandlerTest.java
package org.jeasy.random.validation;

import jakarta.validation.ConstraintViolation;
import jakarta.validation.Validation;
import jakarta.validation.Validator;
import jakarta.validation.ValidatorFactory;
import jakarta.validation.constraints.Future;
import jakarta.validation.constraints.FutureOrPresent;
import org.jeasy.random.EasyRandom;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;

import java.time.*;
import java.time.chrono.HijrahDate;
import java.time.chrono.JapaneseDate;
import java.time.chrono.MinguoDate;
import java.time.chrono.ThaiBuddhistDate;
import java.util.Calendar;
import java.util.Date;
import java.util.Set;

import static org.assertj.core.api.Assertions.assertThat;

class FutureOrPresentAnnotationHandlerTest {

  private EasyRandom easyRandom;

  @BeforeEach
  void setUp() {
    easyRandom = new EasyRandom();
  }

  @Test
  void generatedBeanShouldBeValidAccordingToValidationConstraints() {
    TestBean testBean = easyRandom.nextObject(TestBean.class);

    assertThat(testBean.getTestDate()).isAfterOrEqualTo(Date.from(Instant.now()));
    assertThat(testBean.getTestCalendar()).isGreaterThanOrEqualTo(Calendar.getInstance());
    assertThat(testBean.getTestInstant()).isAfterOrEqualTo(Instant.now());
    assertThat(testBean.getTestLocalDate()).isAfterOrEqualTo(LocalDate.now());
    assertThat(testBean.getTestLocalDateTime()).isAfterOrEqualTo(LocalDateTime.now());
    assertThat(testBean.getTestLocalTime()).isAfterOrEqualTo(LocalTime.now());
    assertThat(testBean.getTestMonthDay()).isGreaterThanOrEqualTo(MonthDay.now());
    assertThat(testBean.getTestOffsetDateTime()).isAfterOrEqualTo(OffsetDateTime.now());
    assertThat(testBean.getTestOffsetTime()).isAfterOrEqualTo(OffsetTime.now());
    assertThat(testBean.getTestYear()).isGreaterThanOrEqualTo(Year.now());
    assertThat(testBean.getTestYearMonth()).isGreaterThanOrEqualTo(YearMonth.now());
    assertThat(testBean.getTestZonedDateTime()).isAfterOrEqualTo(ZonedDateTime.now());
    assertThat(testBean.getTestHijrahDate()).isGreaterThanOrEqualTo(HijrahDate.now());
    assertThat(testBean.getTestJapaneseDate()).isGreaterThanOrEqualTo(JapaneseDate.now());
    assertThat(testBean.getTestMinguoDate()).isGreaterThanOrEqualTo(MinguoDate.now());
    assertThat(testBean.getTestThaiBuddhistDate()).isGreaterThanOrEqualTo(ThaiBuddhistDate.now());
  }

  @Test
  void generatedBeanShouldBeValidUsingBeanValidationApi() {
    TestBean testBean = easyRandom.nextObject(TestBean.class);

    Validator validator;
    try (ValidatorFactory validatorFactory = Validation.buildDefaultValidatorFactory()) {
      validator = validatorFactory.getValidator();
    }
    Set<ConstraintViolation<TestBean>> violations = validator.validate(testBean);

    assertThat(violations).isEmpty();
  }

  static class TestBean {
    @FutureOrPresent private Date testDate;

    public Date getTestDate() {
      return testDate;
    }

    @FutureOrPresent private Calendar testCalendar;

    public Calendar getTestCalendar() {
      return testCalendar;
    }

    @FutureOrPresent private Instant testInstant;

    public Instant getTestInstant() {
      return testInstant;
    }

    @FutureOrPresent private LocalDate testLocalDate;

    public LocalDate getTestLocalDate() {
      return testLocalDate;
    }

    @FutureOrPresent private LocalDateTime testLocalDateTime;

    public LocalDateTime getTestLocalDateTime() {
      return testLocalDateTime;
    }

    @FutureOrPresent private LocalTime testLocalTime;

    public LocalTime getTestLocalTime() {
      return testLocalTime;
    }

    @FutureOrPresent private MonthDay testMonthDay;

    public MonthDay getTestMonthDay() {
      return testMonthDay;
    }

    @FutureOrPresent private OffsetDateTime testOffsetDateTime;

    public OffsetDateTime getTestOffsetDateTime() {
      return testOffsetDateTime;
    }

    @FutureOrPresent private OffsetTime testOffsetTime;

    public OffsetTime getTestOffsetTime() {
      return testOffsetTime;
    }

    @FutureOrPresent private Year testYear;

    public Year getTestYear() {
      return testYear;
    }

    @FutureOrPresent private YearMonth testYearMonth;

    public YearMonth getTestYearMonth() {
      return testYearMonth;
    }

    @FutureOrPresent private ZonedDateTime testZonedDateTime;

    public ZonedDateTime getTestZonedDateTime() {
      return testZonedDateTime;
    }

    @FutureOrPresent private HijrahDate testHijrahDate;

    public HijrahDate getTestHijrahDate() {
      return testHijrahDate;
    }

    @FutureOrPresent private JapaneseDate testJapaneseDate;

    public JapaneseDate getTestJapaneseDate() {
      return testJapaneseDate;
    }

    @FutureOrPresent private MinguoDate testMinguoDate;

    public MinguoDate getTestMinguoDate() {
      return testMinguoDate;
    }

    @FutureOrPresent private ThaiBuddhistDate testThaiBuddhistDate;

    public ThaiBuddhistDate getTestThaiBuddhistDate() {
      return testThaiBuddhistDate;
    }
  }
}

File: bean-validation/src/test/kotlin/io/github/krandom/validation/FutureOrPresentAnnotationHandlerTest.kt
package io.github.krandom.validation

import io.github.krandom.KRandom
import io.kotest.matchers.collections.shouldBeEmpty
import io.kotest.matchers.comparables.shouldBeGreaterThanOrEqualTo
import io.kotest.matchers.nulls.shouldNotBeNull
import jakarta.validation.Validation
import jakarta.validation.Validator
import jakarta.validation.constraints.FutureOrPresent
import java.time.Instant
import java.time.LocalDate
import java.time.LocalDateTime
import java.time.LocalTime
import java.time.MonthDay
import java.time.OffsetDateTime
import java.time.OffsetTime
import java.time.Year
import java.time.YearMonth
import java.time.ZonedDateTime
import java.time.chrono.HijrahDate
import java.time.chrono.JapaneseDate
import java.time.chrono.MinguoDate
import java.time.chrono.ThaiBuddhistDate
import java.util.Calendar
import java.util.Date
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test

internal class FutureOrPresentAnnotationHandlerTest {
  private lateinit var kRandom: KRandom

  @BeforeEach
  fun setUp() {
    kRandom = KRandom()
  }

  @Test
  fun `generated bean should be valid according to validation constraints`() {
    val testBean = kRandom.nextObject(TestBean::class.java)

    testBean.shouldNotBeNull()
    testBean.testDate shouldBeGreaterThanOrEqualTo Date.from(Instant.now())
    testBean.testCalendar shouldBeGreaterThanOrEqualTo Calendar.getInstance()
    testBean.testInstant shouldBeGreaterThanOrEqualTo Instant.now()
    testBean.testLocalDate shouldBeGreaterThanOrEqualTo LocalDate.now()
    testBean.testLocalDateTime shouldBeGreaterThanOrEqualTo LocalDateTime.now()
    testBean.testLocalTime shouldBeGreaterThanOrEqualTo LocalTime.now()
    testBean.testMonthDay shouldBeGreaterThanOrEqualTo MonthDay.now()
    testBean.testOffsetDateTime shouldBeGreaterThanOrEqualTo OffsetDateTime.now()
    testBean.testOffsetTime shouldBeGreaterThanOrEqualTo OffsetTime.now()
    testBean.testYear shouldBeGreaterThanOrEqualTo Year.now()
    testBean.testYearMonth shouldBeGreaterThanOrEqualTo YearMonth.now()
    testBean.testZonedDateTime shouldBeGreaterThanOrEqualTo ZonedDateTime.now()
    testBean.testHijrahDate shouldBeGreaterThanOrEqualTo HijrahDate.now()
    testBean.testJapaneseDate shouldBeGreaterThanOrEqualTo JapaneseDate.now()
    testBean.testMinguoDate shouldBeGreaterThanOrEqualTo MinguoDate.now()
    testBean.testThaiBuddhistDate shouldBeGreaterThanOrEqualTo ThaiBuddhistDate.now()
  }

  @Test
  fun `generated bean should be valid using bean validation api`() {
    val testBean = kRandom.nextObject(TestBean::class.java)

    val validator: Validator
    Validation.buildDefaultValidatorFactory().use { validatorFactory ->
      validator = validatorFactory.validator
    }
    val violations = validator.validate(testBean)

    violations.shouldBeEmpty()
  }

  internal data class TestBean(
    @field:FutureOrPresent val testDate: Date,
    @field:FutureOrPresent val testCalendar: Calendar,
    @field:FutureOrPresent val testInstant: Instant,
    @field:FutureOrPresent val testLocalDate: LocalDate,
    @field:FutureOrPresent val testLocalDateTime: LocalDateTime,
    @field:FutureOrPresent val testLocalTime: LocalTime,
    @field:FutureOrPresent val testMonthDay: MonthDay,
    @field:FutureOrPresent val testOffsetDateTime: OffsetDateTime,
    @field:FutureOrPresent val testOffsetTime: OffsetTime,
    @field:FutureOrPresent val testYear: Year,
    @field:FutureOrPresent val testYearMonth: YearMonth,
    @field:FutureOrPresent val testZonedDateTime: ZonedDateTime,
    @field:FutureOrPresent val testHijrahDate: HijrahDate,
    @field:FutureOrPresent val testJapaneseDate: JapaneseDate,
    @field:FutureOrPresent val testMinguoDate: MinguoDate,
    @field:FutureOrPresent val testThaiBuddhistDate: ThaiBuddhistDate,
  )
}
